Decoupled front end

dhschall opened the pull request:
This PR adds support for a decoupled front-end to gem5. It was extensively tested with x86 and ARM.
The full implementation, as well as an example of how to use it, is described here:
https://github.com/dhschall/gem5-fdp

An example config file to run a simple hello world was added to the config folder: configs/example/gem5_library/fdp-hello.py

dhschall committed:
cpu: Implement decoupled front-end

This commit implements the decoupled front-end (FDP).
Here just the most important bits:

- Creates a new pipeline stage called BAC (Branch and
  Address Calculation).
- The branch predictor is moved to the new stage
  and with it all the functionality related to it.
  (e.g. squashing is done in BAC)
- The BAC stage contains two parts the branch
  predictor part and the next PC calculation. Both
  can be decoupled and connected with the fetch
  target queue.
- New classes for fetch targets and the fetch
  target queue were added. The fetch targets contain
  branch prediction information that gets conveyed to
  the fetch stage. The FTQ holds fetch targets and
  connect the BAC with the fetch stage.

Change-Id: I129ec15f38588bfdfb3eef84c4698e92857354ff
Signed-off-by: David Schall <david.schall@ed.ac.uk>
Signed-off-by: David Schall <david.schall@tum.de>

dhschall committed:
mem-cache: Add fetch directed prefetcher

Adds a new instruction prefetcher that listens to
fetch target queue insertions.
Supports functional as well as timing translations.

Change-Id: I7783501c503563901eb710518d5cb2df0f9c8524
Signed-off-by: David Schall <david.schall@ed.ac.uk>

odanrc commented:
A small note: there is another known fdp in the context of prefetching - the feedback directed prefetcher. It would probably be better to rename the file to not use the abbreviation.

trollet91 commented:
Hello,
I'm starting to evaluate this implementation of decoupled front-end + fdp for research purposes.
First thanks for the excellent work, the code is very clean and readable!

I cherry-picked the commits of your personal branch with the latest fix for fetchTargetWidth in bytes and not instructions and for the prefetcher memory leak.

For a variety of workloads, I don't find significant IPC change. Most of them have a slight IPC decrease while using decoupled front-end over the old front-end. Even for workload that are fronted bounded.
I didn't try the FDP, I want to fully understand the decoupled front-end trade-offs before.

In my configuration, the fetch buffer size is 32 and I tries various size of fetchTargetWidth from 4 to 124. All of them have worse perf than the old front-end:/

I have 2 main questions after reading your code:

1- I see that in your implementation,  the fetch stage can only have 1 concurrent access to the icache. So the only speedup available is using FDP to prefetch entries waiting in the FTQ using the rest of the bandwidth (basically all MSHR - 1)
IIUC, the other implem from @bgodala removes this limits and allow for concurrent icache access. It then reorder the response in a buffer.
What are your thoughts on this?

2- While reading the industry perspective paper I also see one major difference: in the paper, an FTQ entry is created when a taken branch is encountered. IIUC it might takes multiples cycles before a taken branch is detected and an entry is added in the FTQ.
In your implem, each cycle where the BAC stage is not blocked, an entry will be created, even if no branch are detected.
Why did you choose to implement this way?

dhschall commented:
Hi @trollet91 ,
thank you very much for taking the time to look into it and for the great comments.
First of all, I can confirm that we (@rjc-arch and me) also found that the performance is still not good for a bunch of workloads. I guess I made a bug while refactoring the changes and I currently try to find it.

Just for me as context. What architecture (arm/x86) are you using and are you on this branch?
Also, what do you mean by I didn't try the FDP. Do you mean you have `decoupledFrontEnd=False`? 

RE 1: Great question! You are correct in the current implementation I put the prefetching in a separate component that can be connected to the I-cache (`./src/mem/prefetch/fdp.xx`). However, I don't think it's a nice solution as its quite error prone. I like more the implementation of @bgodala and I started to rework that part.

RE 2: My understanding of IV is that every cycle a prediction is made on an entire 16B aligned address block. The first taken branch ends the fetch target or, if no branch is taken, the beginning of the next sequential 32B block boundary. 
However, limiting the maximum fetch target size has another reason. If your BTB is completely cold or you have a very long sequential block you will not feed the FTQ with fetch targets and the fetch unit has to wait. Producing one fetch target per cycle ensures that the fetch unit has to wait only one cycle after a squash to start fetching.

Please let me know if you find bugs or have further questions or suggestions.

atrah22 commented:
Hi @dhschall ,
Any progress on the merging of fdp code as per the implementation in https://github.com/bgodala/gem5/tree/fdip to https://github.com/dhschall/gem5-fdp ?
I assume, the pipeline stage, bac, and how fetch stage access information from bac stage will remain the same. Will there be change in ftq implementation (src/cpu/o3/ftq.cc) ?

dhschall commented:
Hey @atrah22, yes I started to reworked my code with some of the ideas from the other FDP implementation.
It requires some changes especially in the BAC stage.
It works so far but I have to do some code refactoring. Will try to complete this in the next days.

trollet91 commented:
Hello @dhschall 

For now I'm only working on ARM.
Actually my configuration was wrong on the old test.
I had the misunderstanding that fetch was sending multiple requests in paralleled and on top of that the hardware prefetcher that you used was used to prefetch way in advance.

My goal was to only test the fetch part (the fdip hwp was disabled) and it was not giving me improvements. The performance stagnation is actually expected as the decoupled fronted without the prefetch is pretty much similar to the old front-end.

I am re-evaluating the performance with the HWP enabled. But seems like there is still a memory leak as memory explodes during long simulations (even with the patch that you sent in your local branch). Currently trying to fix that.

1. From a quick read of the other fdip implem, the fetch engine is sending multiple fetches in parallel. It reads the response and reorder them before sending the bytes to decode.
I'm not sure this implementation is better. From my understanding front-end have still only one concurrent fetch access to icache, and the rest is prefetched. But I don't have fetch engine uarch knowledge on latest high performance cores.

2. I'm preoccupied by the ease of configuration of the front-end.
Now we have two configurations  variables that may clash .
We have the fetchBufferSize that gives the number of bytes to fetch on each icache access.
We also have fetchTargetWidth that gives the number of bytes to lookup. In the current implem it also gives the maximum size of each FTQ entry (it creates a FTQ entry at the end if it does not find any branch).
What happens on straight line code if fetchTargetWidth is bigger than fetchBufferSize.
Ex: 100000 straight line nops
FetchBufferSize 32
FetchTargetWidth 40
Does this means rgat one cycles the front-end is going to fetch 32 bytes, then 8, then 32, then 8 in the current implem?
If it is the case:
Creating FTQ entries that are multiples of fetchBufferSize could solve this issue without adding complex code, to the already very complex fetch code.
Another solution would be to put basic block in the FTQ. To avoid the problem with cold BTB, we could have a maximum size before creating a FTQ (that is a multiple of fetchBufferWidth).

Would be nice to have other persons with actually uarch knowledge that could share insights (non confidential of course 😉) @andysan maybe some ppl for ARM?

atrah22 commented:
Hello @dhschall , I have question (or maybe it is a bug). 
Btw, I am using a different BTB structure, with different algorithm for lookup and update as per the RTL code (confidential). However, the flow is similar to your default code, in case of a BTBHit, branch predictor (predict function) is called. 
Anyway, the **bug/my question** seems to be regarding the way the FTQ is updated in bac.cc file in **_void BAC::generateFetchTargets(ThreadID tid, bool &status_change)_** function. Below is the part of the code. Please go through my comment lines 

```
cur_pc.set(search_addr);  
//cur_pc is set to either the last instruction addr or the branch instruction addr
..........
.........
if (branch_found) { 

        // branch_found as per the BTB (bpu->BTBValid(tid, search_addr);)
       
         staticInst = bpu->BTBLookupInst(tid, cur_pc);
        assert(staticInst);

        predict_taken = predict(tid, staticInst, curFT, *next_pc);

        DPRINTF(BAC, "[tid:%i, ftn:%llu] Branch found at PC %#x "
                "taken?:%i, target:%#x\n",
                tid, curFT->ftNum(), cur_pc.instAddr(),
                predict_taken, next_pc->instAddr());

        stats.branches++;
        if (predict_taken) {
            stats.predTakenBranches++;
        }
        // if predict_taken == false, then that means the BTB prediction didn't match with the branch predictor

    } else {
        next_pc->set(cur_pc.instAddr() + minInstSize);
    }

   /**
    * The bug/question seems to be regarding the following line, when branch_found == true && predict_taken == false
    * curFT-> finalize function makes the fetchTarget endPC = curPC, branch_found = true, predict_taken = false
    * When updatePC is called from fetch stage and thereby updatePreDecode function is called, it looks for fetchTarget's predictTaken() function which returns predict_taken.
    * So what happens is, it detects that prediction is not taken but a branch is found. However the endPC is still the branch inst addr (cur_pc). The fetch stage will halt fetching the next instruction of the fetchTarget (in the same cycle) and will pop the fetchTarget from FTQ. However, in case a branch is not taken, the fetch stage should be fetching the next instruction of the fetchTarget without any halt and fetchTarget should not be popped from FTQ.
    * In case of branch_found = true && predict_taken=false, shouldn't the cur_pc be equal to the end of cacheline (for my case) or the end of stat_addr + fetchTargetWidth (as per the BTB valid check algorithm in your code) rather than the branch inst PC?
   */
    
    curFT->finalize(cur_pc, curFT->ftNum(), branch_found,
                        predict_taken, *next_pc);
```

dhschall commented:
Hey @atrah22, thanks for the hint.
You are right. In the current implementation fetch stops fetching whether the branch was taken or not.
I have to check how to handle this.
I'm not sure if your fix will solve the problem entirely. The problem is the BAC stage does the same and ends on any branch. If you continue without poping and moving to the next FT then BAC and Fetch are out of sync and the `inRange()` check will fail.
I either have to continue in the BAC stage or allow processing multiple FTs in one cycle.
IMHO the former one is better but then I need to support multiple branches to be in one FT where only one can be taken.

atrah22 commented:
Hello @dhschall , Yes I agree that the support for multiple branches to be in one FT where only one can be taken is better.

For now though, I modified only fetch Stage in fetch.cc (no change in bac.cc file) to allow processing multiple FTs in one cycle when there is a not taken branch and the next PC is in the following FT. It shows performance improvement for a FS simulation of relatively big single core workload.
This feature could be important for performance.

trollet91 commented:
I think that we need to define clearly how we generate FT from the BAC stage.

Things that we need to agree on and clearly define:
- How many FT can be generated per cycles?
What is the size of a FT entry?
- Does FT entries cross cache lines?
- Do we add alignment constrains to FT entries?
- How may taken branch (cond and/or uncond) branches does the BAC can handle in one cycle?
- How many concurrent fetch access in the fetch stage?
- How many concurrent prefetch access in the fetch stage?
- Can the frontend send fetch and prefetch during the same cycle?



I think that the we need to start with a simple FT configuration and then we can try to add more complex behaviors later on.

BAC stage 
- configuration:
    - FetchTargetWidth: how many bytes scanned each cycles
   - NumBranchTaken: max number of taken branches in one cycle
        - NumUnconditionalTaken: max number of unconditional taken per cycle
        - NumConditionalTaken: max number of conditional taken per cycles
	For ex I read that (never reversed it myself) that some processors handles 2 taken unconditional branches per cycles. And with pipeline width soon reaching 10 (currently 9 in Apple A17), it's a matter of time before we get a core that that handles 2 conditional taken branches (if it's not already the case).
    - NumberFT: max number of FT that can be generated in one cycle. For exemple in the case where we handle multiple taken branches per cycle, we need to generate multiples FT (Or  complexify FT to handle 2 non contiguous fetch in one entry, but I don't think it's a good idea)
			 
- BAC generates FT when:
    - taken branch detected
    - reach end of cacheline
    - reach size of fetchBufferSize
This allows the fetch stage to fully consume entries in 1 go and will simplify the consume logic while not making the produce logic much more complex. IMO this would be a good first implementation.

Later on we can add more complex behavior like cross-cacheline FT, alignments, ...


Fetch stage
- configuration:
    - NumberConcurrentFetch: number of concurrent fetch access allowed using consecutive FTQ entries. We have to define exactly the behavior as we need to model a small buffer that will be used to temporarily keep the data if the responses are out of order.
    - NumberConcurrentPrefetch: number of concurrent prefetch access allowed. Do we need this constrain? Because MSHR are going to be a natural bottleneck for the prefetches.
    - Configure the access ports to icache. Do we allow 1 prefetch and 1 fetch the same cycle? 2 fetch same cycle? 2 prefetch same cycle? 2 fetch and 2 prefetch same cycle? â€¦ Need to think about a convenient way to express it.
	Each cycle the fetch stage consumes entries for fetch/prefetch.



Tell me if that would be acceptable for you? Do you need additional behavior?
I'm starting to change the behavior of the BAC (in David's pf-rework branch) to implement what I just explained for the BAC.

dhschall commented:
Thanks @trollet91 for coming up with this list! All points look quite reasonable to me and would improve the implementation.

Some comments on your points:
- #FT: I agree to start with one FT per branch (taken or not). It might look weird as all references say its one FT per cycle but the problem is the branch predictor history and how to manage that.

- FT cross CL/alignment: For prefetching and translation it might be better to not cross. The problem is in x86 instructions can cross cache lines

- #taken branches: Might be easier to start with one but in case we can have more would be great. After all its a simulator to explore future design decisions not only simulating SOTA.

- Max fetch target size: Limiting it to `fetchBufferSize` makes only sense if we generate several fetch targets in one cycle. The BAC stage bandwidth must be larger than the fetch stage. Otherwise, the BAC stage can never gain sufficient prefetch distance.

- Why do you want concurrent fetches and in addition prefetches? I understand it that the demand read into the fetch buffer is a `fetch` all others are prefetches. Of course, you can say you pipeline the next already but what is the difference to a prefetch?

I look forward to your modifications :) 
Thanks again!

trollet91 commented:
> #FT: I agree to start with one FT per branch (taken or not). It might look weird as all references say its one FT per cycle but the problem is the branch predictor history and how to manage that.

Why should we create a FT on a detected but  non taken branch? Is this to simplify the history management?

> Max fetch target size: Limiting it to fetchBufferSize makes only sense if we generate several fetch targets in one cycle. The BAC stage bandwidth must be larger than the fetch stage. Otherwise, the BAC stage can never gain sufficient prefetch distance.

I think we need to be able to generate multiple FT in one cycle to support multiples taken branches in one cycle. The other solution is to have two non contiguous part in one FT. This is harder to implem and I doubt that it will be implemented like that in a real design.

Regarding the FT creation: let's say that fetchBufferSize is 32, and fetchTargetWidth is 48.
We have 3 options:
1/ Generate a FT for 48 bytes. This means the front-end as to fetch multiple times from the same FT.
2/ Generate a FT of 32 bytes and another FT of 16 bytes.
3/ generate one FT of 32 bytes. The 16 bytes leftover will be included in the next FT created the cycle after.

IMO the design  3/ is is easier to implement than 1/ and 2/ because the fetch stage will fully consume entry at once. This simplifies fetch logic. The generation logic is now slightly more complex but should stay very simple.

FT creation policy:
- taken branch detected
- reach end of cacheline
- reach size of fetchBufferSize

Would you agree with a design like that?

> Why do you want concurrent fetches and in addition prefetches?

For a bigger core that predicts 2 taken branches per cycle, you may want to consume two entries per cycle. But we can discuss that later it's not important now.

atrah22 commented:
Hello, @dhschall and @trollet91, Full system simulation even for a small workload keeps running (stuck in loop) I think due to the warning message (in performCacheAccess function in fetch.cc) . Address is outside of physical memory, stopping fetch. This small workload (CCa of microbench) is very simple and worked always for other branch (fdp) or any version of gem5. I think, there may be a bug in pf-rework branch.
```
if (!cpu->system->isMemAddr(mem_req->getPaddr())) {
        warn("Address %#x is outside of physical memory, stopping fetch\n",
                mem_req->getPaddr());
        fetchStatus[tid] = NoGoodAddr;
        memReq[tid] = NULL;
        return false;
    }

````

dhschall commented:
Hi @atrah22 , thanks for letting me know! 
That's very likely. The last commit was just a first attempt to rework and integrate suggestions from @andysan and the other FDP implementation. Though, I did not have the time to do proper cleanup and testing.

atrah22 commented:
> Hi @atrah22 , thanks for letting me know! That's very likely. The last commit was just a first attempt to rework and integrate suggestions from @andysan and the other FDP implementation. Though, I did not have the time to do proper cleanup and testing.

Hello, just for record, I found another issue regarding tage predictor when the "speculativeHistUpdate" flag is turned to false. There is a segmentation fault. Maybe its due to the commit "cpu: BPU support for surprise branches"

trollet91 commented:
Update for anyone interested in this feature.
I took over the code from David Schall. My goal is to make the implem a bit more close to a modern processor.

The biggest change is that the Fetch class is now pure virtual and inherited by two classes 'SimpleFetch' and 'DecoupledFetch'.
SimpleFetch is the old implem and DecoupleFetch is a total new fetch pipeline implem.
Having the decoupled fetch logic on top of the old fetch logic was too hard to write/validate.

Right now I'm using the new decoupled frontend in ARM SE for an internal research project.
It's already bringing performance improvements compare to the old frontend in benchmarks with code footprint too big for icache. And similar performance for small benchmarks that fits in cache.
Once I'll be totally confident about the behavior, I will add back the code from David to handle x86 jump in uops and FS specific stalls and finally send the code to review. But it will still take some time.

atrah22 commented:
@trollet91,
Is it possible now to look at the code from your git branch for the decoupleFetch even if not ready for review? That would be great. thanks.

trollet91 commented:
@atrah22 Sorry right now it's in a internal zone of the company to run internal benchmarks that stresses the icache, AKA not SPEC 😄 . Moving it to upstream will require some paperwork so I will only do it once.

atrah22 commented:
@dhschall , @trollet91 ,
I have another question,
After simulation, the stat, _bac.noBranchMisspredict_ is 3,000 ~ 4,000 for a micro-benchmark. Why a non-control instruction causes a squash/flush and why a non-control instruction is miss predicted? The DPRINTF comment "Squashing due to mispredict of non-control instruction" is in the bac.cc file.

dhschall committed:
arch-x86: Add code hit/miss stats to TLB

The current implementation only distingnish
between read and write misses & hits. This
change adds also statistics for code.

dhschall committed:
cpu: Add FDP support for TAGE-SC-L.

In order to work with FDP TAGE-SC-L must implement
the branchPlaceholder function.

dhschall commented:
I want to revive the discussion of this PR. All pre required commits are now merged so we can start looking into how to get this on in.
I'm happy to take on any feedback to improve it or modularize it.

Todo's:

- [x] Add an example setup in the gem5 configs. (maybe a board with FDP support)
- [ ] Think about if we could make the fetch unit a virtual base class with a coupled/decoupled class that overrides it.

dhschall committed:
misc: Example config file for FDP

erin-le commented:
@dhschall Hello, thanks for the update! We'll try to get someone to review this.

rjc-arch commented:
Hi @erin-le. 
@dhschall and I are testing and reviewing this PR at the moment. 
My understanding is that @dhschall would welcome input and reviews from everyone, but could we hold off merging the PR until we have finished fully testing it for expected behaviour and performance uplift?
I will also do a full code review shortly too.

dhschall commented:
Hi @erin-le, yes @rjc-arch will help me with more extensive testing. We will also report some of the expected performance differences when enabling the decoupling.
But in the meantime I feel it would be good to get some high level feedback from at least one more experienced developer.

powerjg commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+
+
+cache_hierarchy = CacheHierarchy(
+    l1i_size="32KiB", l1d_size="32KiB", l2_size="1MB"
```
```suggestion
    l1i_size="32KiB", l1d_size="32KiB", l2_size="1MiB"
```

powerjg commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+        4  # Note Arm has a 2 byte thumb mode but we ignore it here
+    )
+elif args.isa == "RiscV":
+    cpu.minInstSize = 4  # RiscV has a 2 byte compressed mode.
```
IIRC, compressed mode is enabled by default. Should this be 2?

powerjg commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+## Branch Predictor Unit --------------------------------
+# The decoupled front-end relies on the BTB to discover
+# branches and uses `takenOnlyHistory`.
+# Also at the moment only LTAGE and TAGE-SC-L are supported.
```
Why are these the only branch predictors supported? (I'm mostly just curious what the limitations are)

dhschall replied:
The branch predictors must implement the `branchPlaceholder` method which is trivial for simple BPUs. I've added now a note

powerjg commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+
+
+class BPU(LTAGE):
+    instShiftAmt = 2 if args.isa == "Arm" else 0
```
Could you add a comment here explaining why this is 2 for Arm? E.g., I would expect that it would also be 2 for RISC-V given the code above that RISC-V is assuming 4-byte instructions.

dhschall replied:
Yep

powerjg commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+print(
+    "Running {} on {}, FDP {}".format(
+        args.workload, args.isa, "disabled" if args.disable_fdp else "enabled"
+    )
```
Nit: I prefer f-strings.

```suggestion
    f"Running {args.workload} on {args.isa}, FDP {'disabled' if args.disable_fdp else 'enabled'}".
```

powerjg commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+# program compiled to the ARM ISA. The `Resource` class will automatically
+# download the binary from the gem5 Resources cloud bucket if it's not already
+# present.
+board.set_se_binary_workload(
+    # The `Resource` class reads the `resources.json` file from the gem5
+    # resources repository:
+    # https://gem5.googlesource.com/public/gem5-resource.
+    # Any resource specified in this file will be automatically retrieved.
+    # At the time of writing, this file is a WIP and does not contain all
+    # resources. Jira ticket: https://gem5.atlassian.net/browse/GEM5-1096
+    obtain_resource(workloads[args.workload][args.isa])
+)
```
You should be able to use just `set_workload`

```suggestion
board.set_workload(obtain_resource(workloads[args.workload][args.isa]))
```

dhschall replied:
It tells me `AttributeError: 'BinaryResource' object has no attribute 'get_function_str'`

powerjg replied:
My bad. I thought we had workloads for these. @Harshil2107 any chance you want to make some workloads for the following binaries? Really, all of the hello binaries should have workloads.

```
workloads = {
    "hello": {
        "Arm": "arm-hello64-static",
        "X86": "x86-hello64-static",
        "RISCV": "riscv-hello",
    },
}
```

powerjg commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+print(
+    "Exiting @ tick {} because {}.".format(
+        simulator.get_current_tick(), simulator.get_last_exit_event_cause()
+    )
```
```suggestion
    f"Exiting @ tick {simulator.get_current_tick()} because {simulator.get_last_exit_event_cause()}.".
```

Though, to be honest, this is a pretty useless print statement. I've been removing these from new scripts, personally.

powerjg commented on src/arch/x86/decoder.cc:
```diff
@@ -696,7 +696,10 @@ Decoder::decode(PCStateBase &next_pc)
-    return microcodeRom.fetchMicroop(micropc, curMacroop);
+  auto si = microcodeRom.fetchMicroop(micropc, curMacroop);
+  si->size(4);
+  return si;
```
2 spaces, not 4

powerjg replied:
Why this change? I can guess, butit needs a comment.

powerjg commented on src/arch/x86/decoder.cc:
```diff
@@ -696,7 +696,10 @@ Decoder::decode(PCStateBase &next_pc)
+  auto si = microcodeRom.fetchMicroop(micropc, curMacroop);
+  si->size(4);
+  return si;
+    // return microcodeRom.fetchMicroop(micropc, curMacroop);
```
Remove commented code.

powerjg commented on src/arch/x86/tlb.cc:
```diff
@@ -418,19 +418,29 @@ TLB::translate(const RequestPtr &req,
+                case BaseMMU::Execute:
+                    stats.exAccesses++; break;
+                default:
+                    break;
```
Default should probably be a panic.

powerjg commented on src/arch/x86/tlb.cc:
```diff
@@ -418,19 +418,29 @@ TLB::translate(const RequestPtr &req,
+                    case BaseMMU::Execute:
+                        stats.exMisses++; break;
+                    default:
+                        break;
```
Ditto

powerjg commented on src/cpu/o3/BaseO3CPU.py:
```diff
@@ -217,3 +223,17 @@ def support_take_over(cls):
     recvRespBufferSize = Param.Unsigned(
         64, "Maximum number of receive response bytes per cycle"
     )
+    numFTQEntries = Param.Unsigned(
```
Could you add to the help message here and the other options that are only used if the decoupled front end is enabled?

powerjg commented on src/cpu/o3/SConscript:
```diff
@@ -1,5 +1,17 @@
+# licensed hereunder.  You may use the software subject to the license
+# terms below provided that you ensure that this notice is replicated
+# unmodified and in its entirety in all distributions of the software,
+# modified or unmodified, in source code or in binary form.
```
Do you have to add this special extension to the license? It would be much preferable to not change the license. (It also seems a bit silly to think that there is IP in the sconscript file.)

powerjg commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1069 @@
+    fatal_if(decoupledFrontEnd && (fetchTargetWidth < params.fetchBufferSize),
+            "Fetch target width should be larger than fetch buffer size!");
+
+    for (int i = 0; i < MaxThreads; i++) {
```
Has this been tested with multiple threads? If not, I suggest putting a `fatal_if(numThreads > 1, "More than 1 thread has not been tested with the decoupled front end")`

powerjg commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1069 @@
+        stalls[i] = {false, false, false};
+    }
+
+    assert(bpu!=nullptr);
```
```suggestion
    assert(bpu != nullptr);
```

This PR may be a good test for the clang format...

dhschall replied:
I run the clang formatter now on it.

powerjg commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1069 @@
+}
+
+std::string
+BAC::name() const {
```
```suggestion
BAC::name() const
{
```

powerjg commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1069 @@
+void
+BAC::setTimeBuffer(TimeBuffer<TimeStruct> *time_buffer)
+{
+    // TODO:
```
TODO what?

powerjg commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1069 @@
+
+
+
+
```
```suggestion
```

powerjg commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1069 @@
+}
+
+
+
```
Lots of extra whitespace... I'll stop commenting on style now.

dhschall replied:
Yes, used now the clang formatter to format the entire file

powerjg commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1069 @@
+    std::list<ThreadID>::iterator threads = activeThreads->begin();
+    std::list<ThreadID>::iterator end = activeThreads->end();
+
+if (decoupledFrontEnd) {
```
(maybe) the last style comment: This indentation is wrong (and confusing).

powerjg commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1069 @@
+        ThreadID tid = *threads++;
+
+        // Check stall and squash signals first.
+        status_change |= checkSignalsAndUpdate(tid);
```
Is this logical or bitwise or... I think it's supposed to be logical. I would make that explicit.

```suggestion
        status_change = status_change || checkSignalsAndUpdate(tid);
```

powerjg commented on src/cpu/o3/bac.hh:
```diff
@@ -0,0 +1,470 @@
+ * Copyright (c) 2022-2023 The University of Edinburgh
+ * All rights reserved
+ *
+ * The license below extends only to copyright in the software and shall
+ * not be construed as granting a license to any other intellectual
+ * property including but not limited to intellectual property relating
+ * to a hardware implementation of the functionality of the software
+ * licensed hereunder.  You may use the software subject to the license
+ * terms below provided that you ensure that this notice is replicated
+ * unmodified and in its entirety in all distributions of the software,
+ * modified or unmodified, in source code or in binary form.
```
Did University of Edinburgh say that you had to have this license modification? If not, I would prefer to drop it from all files.

rjc-arch replied:
Hi @powerjg. I believe this work was partly funded by Arm Ltd., and is therefore following the standard format for gem5 copyright notices used by Arm.

powerjg commented on src/cpu/o3/fetch.cc:
```diff
@@ -1087,16 +1071,39 @@ Fetch::fetch(bool &status_change)
 
+    FetchTargetPtr curFT = ftq->readHead(tid);
+
+    if (decoupledFrontEnd) { // #ifdef FDIP
```
```suggestion
    if (decoupledFrontEnd) {
```

powerjg commented on src/cpu/o3/fetch.cc:
```diff
@@ -1361,6 +1407,8 @@ ThreadID
 {
     if (numThreads > 1) {
+        // More that one thread is not tested with decoupled Frontend.
+        assert(!decoupledFrontEnd);
```
This should probably be a fatal (user configuration error), and it would be good to give the user some information. It is OK if it's here, but it would probably be better in the constructor of the BAC as commented above.

dhschall replied:
Moved the check to the constructor

powerjg commented on src/cpu/o3/ftq.cc:
```diff
@@ -0,0 +1,319 @@
+}
+
+
+/** Fetch Target Qeue Methods ----------------------------- */
```
```suggestion
/** Fetch Target Queue Methods ----------------------------- */
```

powerjg commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,308 @@
+  public:
+    /** Ancore point to attach a branch predictor history.
+     * Will carry information while FT is waiting in th FTQ. */
+    void* bpu_history;
```
Why is this a `void*`?

dhschall replied:
Good point! I'll see if I can make it a normal pointer to the branch predictor history object

powerjg commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,308 @@
+
+
+
+public:
```
```suggestion
  public:
```

powerjg commented on src/cpu/pred/bpred_unit.hh:
```diff
@@ -373,7 +379,7 @@ class BPredUnit : public SimObject
 
         /** The sequence number for the predictor history entry. */
-        const InstSeqNum seqNum;
+        InstSeqNum seqNum;
```
WHy no longer const?

dhschall replied:
Because the decoupled branch predictor makes predictions for branches **before** the dynamic instruction instances are created. Thus, the sequence number is not known.
Once the sequence number is known in the fetch stage the PredictorHistory is updated accordingly.

powerjg commented on src/cpu/pred/tage_sc_l.cc:
```diff
@@ -392,6 +392,17 @@ TAGE_SC_L_TAGE::extraAltCalc(TAGEBase::BranchInfo* bi)
+TAGE_SC_L::branchPlaceholder(ThreadID tid, Addr pc,
+                                bool uncond, void * &bp_history)
+{
+    TageSCLBranchInfo *bi = new TageSCLBranchInfo(*tage,
```
Do we have (or need) a `BaseBranchInfo` to communicate other information from the branch predictor to the FTQ?

What information needs to be communicated? I don't quite understand how this code is used, so it's hard to suggest a better implementation. However, the current implementation seems brittle.

dhschall replied:
The branch placeholder function is to create a history object for branches that were not detected by the branch predictor. E.g. Never taken branches or first time branched. Basically all which are not in the BTB.

This function is called by the fetch unit for all such branches to create a such an history object which is needed in case of a misprediction to role back or squash the history

powerjg commented on src/mem/cache/base.cc:
```diff
@@ -129,8 +129,10 @@ BaseCache::BaseCache(const BaseCacheParams &p, unsigned blk_size)
-    if (prefetcher)
+    if (prefetcher) {
         prefetcher->setParentInfo(system, getProbeManager(), getBlockSize());
+        prefetcher->setCache(this);
```
How is this going to work with Ruby?

rjc-arch replied:
This should probably be re-worked to use the `CacheAccessor` interface introduced in this commit: [mem-cache: decoupled prefetchers from cache](https://github.com/gem5/gem5/commit/af2ee0db306cda5f6cea27f631fc67d87e5fa1c1).

The `BaseCache*` was removed from the `Base` prefetcher in that commit, and replaced with an interface to enable the prefetcher models to be used with Ruby (addressing @powerjg's comment).

powerjg commented on src/mem/cache/prefetch/SConscript:
```diff
@@ -1,5 +1,17 @@
+# Copyright (c) 2023 The University of Edinburgh
+# All rights reserved.
+#
+# The license below extends only to copyright in the software and shall
+# not be construed as granting a license to any other intellectual
+# property including but not limited to intellectual property relating
+# to a hardware implementation of the functionality of the software
+# licensed hereunder.  You may use the software subject to the license
+# terms below provided that you ensure that this notice is replicated
+# unmodified and in its entirety in all distributions of the software,
+# modified or unmodified, in source code or in binary form.
```
Needed?

dhschall replied:
Removed it

powerjg commented on src/mem/cache/prefetch/fdp.hh:
```diff
@@ -0,0 +1,209 @@
+
+    /** Base class overrides */
+    void regProbeListeners() override;
+    void setCache(BaseCache *_cache) override { cache = _cache; }
```
I could be missing something, but I don't see where `_cache` is used. Could we drop this change?

powerjg reviewed (commented):
I didn't review the implementation of the decoupled front end and FDP in detail, but I went through everything else. As long as things perform as expected, I'm happy to merge. Since this is a *new* option, I think that even the performance doesn't have to be perfect on the first merge. 

One thing we should check is that when the decoupled front end is turned off the performance should be about the same as before.

rjc-arch commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,308 @@
+#include "cpu/o3/limits.hh"
+#include "sim/probe/probe.hh"
+
+#define FDIP
```
Looks like this isn't used any more. (The only other reference is commented out.)

rjc-arch commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,308 @@
+typedef InstSeqNum FTSeqNum;
+
+
+/** The fetch target class. */
```
A more descriptive documentation comment might be helpful. (This one just says the same things as the next line.)
You could explain briefly what it represents in the FDP system.

dhschall replied:
Is that ok?

rjc-arch commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,308 @@
+    bool taken;
+
+  public:
+    /** Ancore point to attach a branch predictor history.
```
Typo: Ancor -> Anchor

rjc-arch commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,308 @@
+    const PCStateBase &readStartPC() { return *startPC; }
+
+    /** Read the exit/end address PC */
+    const PCStateBase &readEndPC() { return *endPC; }
```
Should this handle the `endPC == nullptr` case? There doesn't seem to be a separate way for client code to check this dereference will be safe.

rjc-arch commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,308 @@
+                  bool pred_taken, const PCStateBase &pred_pc);
+
+    /** Print the fetch target for debugging. */
+    std::string print();
```
Nit: this should probably be named something like `to_string`, `as_string`, etc.
It doesn't actually do the printing.

rjc-arch commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,308 @@
+    };
+
+    /** Per-thread FTQ status. */
+    Status ftqStatus[MaxThreads];
```
Nothing wrong with this, but it might be worth considering using a `std::array` instead. (Likewise for some other similar instances.)

rjc-arch commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,308 @@
+    /** Number of fetch targets in the FTQ. (per thread) */
+    const unsigned numEntries;
+
+    /** Probe points to attach the FDP perfetcher. */
```
Type: perfetcher -> prefetcher

rjc-arch commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,308 @@
+    /** Returns the size of the ftq for a specific partition*/
+    unsigned size(ThreadID tid);
+
+    /** Returns if a specific thread's queue is full. */
```
"Returns whether or not a specific thread's queue is full." might be slightly less ambiguous wording.

rjc-arch commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,308 @@
+     * Requires squash to recover. */
+    void invalidate(ThreadID tid);
+
+    /** Returns if the FTQ is in a val;id state and its save to consmume
```
Typo: val;id -> valid
Typo: save -> safe

rjc-arch commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,308 @@
+    /** Squashes all fetch targets in the FTQ for a specific thread. */
+    void squash(ThreadID tid);
+
+    /***/
```
Missing documentation comment here.

rjc-arch commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,308 @@
+    /** Updates the head fetch target once its fully processed
+     * In case there is still a branch history attached to the
+     * head fetch target the FTQ goes into invalid state.
+     * @return Wheather or not the update was successful.
```
Typo: Wheather -> Whether

rjc-arch commented on src/cpu/o3/ftq.cc:
```diff
@@ -0,0 +1,319 @@
+
+
+void
+FetchTarget::finalize(const PCStateBase &exit_pc, InstSeqNum sn,
```
The `sn` parameter doesn't seem to be used here. Could it be removed?

rjc-arch commented on src/cpu/o3/ftq.cc:
```diff
@@ -0,0 +1,319 @@
+void
+FTQ::resetState()
+{
+    for (ThreadID tid = 0; tid  < numThreads; tid++) {
```
Would it be better to initialise up to MaxThreads (perhaps with a different status for `tid >= numThreads`). In practice (with correct usage) it doesn't make a difference, but it feels better not to leave any of the array elements uninitialised.

dhschall replied:
I see your point. However, I implemeted it in analogy to the other stages and I think it should be fixed in another PR

rjc-arch commented on src/cpu/o3/ftq.cc:
```diff
@@ -0,0 +1,319 @@
+}
+
+bool
+FTQ::isEmpty() const
```
What is this method used for? It seems like it is querying the 'unused' portion of the FTQ.

dhschall replied:
True, this is useless and also incorrect! Thanks for the catch

rjc-arch commented on src/cpu/o3/ftq.cc:
```diff
@@ -0,0 +1,319 @@
+void
+FTQ::invalidate(ThreadID tid)
+{
+    /** Only a full ftq can be invalid*/
```
Typo: invalid -> invalidated ?
Also the comment says 'full' but the test is for 'not empty'. Which is correct?

rjc-arch commented on src/cpu/o3/ftq.cc:
```diff
@@ -0,0 +1,319 @@
+}
+
+bool
+FTQ::isValid(ThreadID tid)
```
Nit: this seems like slightly strange naming (though I can understand why from a practical point of view). `Valid` is an explicit state in the `Status` enum, but this method will return true for other states like `Full` and `Locked`, which might be a little confusing. Could the `Valid` state or this method be re-named (e.g. `isReady()`, `isAvailable()`, etc)?

rjc-arch commented on src/cpu/o3/ftq.cc:
```diff
@@ -0,0 +1,319 @@
+
+    // TODO make this more efficient
+    // Once the head of the FTQ gets updated and
+    // the FTQ got blocked by a complex instruction resteere
```
Typo: resteere -> resteer

rjc-arch commented on src/cpu/o3/ftq.cc:
```diff
@@ -0,0 +1,319 @@
+bool
+FTQ::updateHead(ThreadID tid)
+{
+    if (ftq[tid].front()->bpu_history != nullptr) {
```
Should there be a check for an empty `ftq` here? Dereferencing `ftq[tid].front()` is undefined behaviour.

rjc-arch commented on src/cpu/o3/ftq.cc:
```diff
@@ -0,0 +1,319 @@
+{
+    occupancy
+        .init(/* base value */ 0,
+              /* last value */ ftq->numEntries,
```
This is safe as it is now, but might be fragile to future code code changes (as you are passing in the `this` pointer of a partially constructed object and then dereferencing it before construction is complete). You could consider passing in the `numEntries` value directly instead.

rjc-arch commented on src/cpu/o3/ftq.cc:
```diff
@@ -0,0 +1,319 @@
+
+
+void
+FTQ::insert(ThreadID tid, FetchTargetPtr fetchTarget)
```
Insert doesn't do any checks to see if the `ftq` is full. 
Would it be better to check internally, rather than rely on the client code to check?

dhschall replied:
I actually added an assert internally already, but good catch!

rjc-arch replied:
Thanks!

rjc-arch commented on src/cpu/o3/ftq.cc:
```diff
@@ -0,0 +1,319 @@
+
+
+bool
+FTQ::updateHead(ThreadID tid)
```
This method looks like it removes/pops the head of the `ftq`, rather than modifying the head element of the queue. Renaming the method to reflect this might make the purpose a bit less ambiguous.

In general, it might be clearer if the method names more closely followed the naming conventions for queues (e.g. `push`, `popHead`, etc.).

dhschall replied:
Well I though more that we update the head pointer.
But I can see what you mean and why that makes sense.

rjc-arch commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,308 @@
+    ProbePointArg<FetchTargetPtr> *ppFTQRemove;
+
+    /** FTQ List of Fetch targets */
+    std::list<FetchTargetPtr> ftq[MaxThreads];
```
Likewise, this could be a `std::array` if you like.

rjc-arch commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+starting point for the FDP implementation.
+As workload a simple "Hello World!" program is used.
+
+FDP is tested with the X86, Arm, RiscV isa which can be specified by
```
This should be written as `RISC-V` (according to the RISC-V [Brand Guidelines](https://riscv.org/about/brand-guidelines/)).

Typo: "which can be specified by with the `--isa` flag" -> "which can be specified with the `--isa` flag" or "which can be specified using the `--isa` flag"

rjc-arch commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+    description="An example configuration script to run FDP."
+)
+
+# The only positional argument accepted is the benchmark name in this script.
```
Is this comment correct? It doesn't look like there are any positional arguments at the moment.

rjc-arch commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+isa_choices = {
+    "X86": ISA.X86,
+    "Arm": ISA.ARM,
+    "RiscV": ISA.RISCV,
```
Likewise, the key name could probably be closer to the brand guidelines.

rjc-arch commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+parser.add_argument(
+    "--disable-fdp",
+    action="store_true",
+    help="Disable FDP to get evaluate baseline",
```
Typo: "Disable FDP to get evaluate baseline" -> "Disable FDP to evaluate baseline"

rjc-arch commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+memory = SingleChannelDDR3_1600(size="32MB")
+
+
+# 1. Instruction prefetcher ---------------------------------------------
```
It might be helpful to have a brief list of the steps needed to instantiate, connect, and enable the FDP here (to save having to find the exact places in the code).

rjc-arch commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+)
+
+
+# 1. Decoupled front-end ------------------------------------------------
```
Typo: "1." -> "2."

rjc-arch commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+
+
+# 1. Decoupled front-end ------------------------------------------------
+# First setup the decoupled front-end. Its implemented in the O3 core.
```
This looks like the "Second"/"Next" step.

rjc-arch commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+# The decoupled front-end leverages the BTB to find branches in the fetch
+# stream. Starting from the end of the last fetch target it will search
+# all addresses until a hit. However, for fixed size instruction architectures
+# like ARM only every n-th address must be checked.
```
Typo: ARM -> Arm

rjc-arch commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+)
+
+
+# The gem5 library simble board which can be used to run simple SE-mode
```
Typo: simble -> simple

rjc-arch commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+)
+
+# Here we set the workload. In this case we want to run a simple "Hello World!"
+# program compiled to the ARM ISA. The `Resource` class will automatically
```
Should this be "compiled to the specified ISA"?

rjc-arch commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+    def __init__(self, l1i_size, l1d_size, l2_size):
+        super().__init__(l1i_size, l1d_size, l2_size)
+
+    def incorporate_cache(self, board: AbstractBoard) -> None:
```
Optional suggestion: could this be re-worked to just iterate over the cores present and add a private cache hierarchy and prefetchers to each core? 

The current implementation assumes a single single-core CPU (which is fine in the context of this example, as the number of CPUs is hard coded to one below), but the code might be slightly clearer if re-worked to support any number of cores in this method.

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+      fetchTargetWidth(params.fetchTargetWidth),
+      minInstSize(params.minInstSize),
+      numThreads(params.numThreads),
+      stats(_cpu,this)
```
Style: needs an extra space after the comma.

rjc-arch commented on src/cpu/o3/bac.hh:
```diff
@@ -0,0 +1,470 @@
+    /** Sets pointer to list of active threads. */
+    void setActiveThreads(std::list<ThreadID> *at_ptr);
+
+    /** Initialize stage. */
```
Copy-and-paste of the wrong documentation comment?

rjc-arch commented on src/cpu/o3/bac.hh:
```diff
@@ -0,0 +1,470 @@
+     *
+     * For determining next PC addresses the BPU relies solely on the BTB.
+     *
+     * The follwing functions are only used in the decoupled scenario.
```
Typo: follwing -> following

rjc-arch commented on src/cpu/o3/bac.hh:
```diff
@@ -0,0 +1,470 @@
+     * with the currently pre-decoded instruction.
+     * The BAC stage will then check if the instruction is a branch and
+     * either preform the branch prediction (non-decoupled scenario) or
+     * read the branch prediction from the currentently processed fetch
```
Typo: currentently -> currently

dhschall commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1069 @@
+{
+    if (!decoupledFrontEnd) return;
+
+    DPRINTF(BAC, "%s(tid:%i): FTQ sz: %i\n", tid, __func__, ftq->size(tid));
```
```suggestion
    DPRINTF(BAC, "%s(tid:%i): FTQ sz: %i\n",  __func__, tid, ftq->size(tid));
```

rjc-arch commented on src/cpu/o3/bac.hh:
```diff
@@ -0,0 +1,470 @@
+ *     fetch stage. Fetch calls the updatePC() method after each
+ *     predecoded instruction to advance the PC to the next instruction.
+ *
+ * Note that this organization means that the BAC stage has spans over
```
Typo: "has spans over" -> "spans"

rjc-arch commented on src/cpu/o3/bac.hh:
```diff
@@ -0,0 +1,470 @@
+
+    void deactivateThread(ThreadID tid) {};
+
+    /** Processing all input signals and create the next fetch target. */
```
Typo: Processing -> Process

rjc-arch commented on src/cpu/o3/bac.hh:
```diff
@@ -0,0 +1,470 @@
+    void updateBACStatus();
+
+    /** Checks all input signals and updates the status as necessary.
+     *  @return: Returns if the status has changed due to input signals.
```
Suggestion: "Returns whether the status..." or "Returns true if the status..." might be a little clearer.

rjc-arch commented on src/cpu/o3/bac.hh:
```diff
@@ -0,0 +1,470 @@
+     * To update the PC the fetch stage will call the updatePC() method
+     * with the currently pre-decoded instruction.
+     * The BAC stage will then check if the instruction is a branch and
+     * either preform the branch prediction (non-decoupled scenario) or
```
Typo: preform -> perform

rjc-arch commented on src/cpu/o3/bac.hh:
```diff
@@ -0,0 +1,470 @@
+      statistics::Scalar branchMisspredict;
+      statistics::Scalar noBranchMisspredict;
+
+      /** Stat for total number of misspredicted instructions. */
```
Copy and paste error in the comment?

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+    stalls[tid].drain = false;
+    stalls[tid].bpu = false;
+
+    assert(ftq!=nullptr);
```
Style: needs spaces around the `!=`.

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+void
+BAC::resetStage()
+{
+    // Setup PC and nextPC with initial state.
```
Suggestion: Could some of this code be deferred to `BAC::clearStates(ThreadID tid)`? 
The `ftq->resetState()` call would be repeated unnecessarily, but otherwise the code is the same. (The repetition could also be avoided with a `tid` specific variant of `FTQ::resetState()`.)

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+        if (!ftq->isEmpty(i))
+            return false;
+
+        // Return false if not idle or drain stalled
```
The comment here doesn't match the code.

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+    bool ret_val = false;
+
+    if (stalls[tid].fetch) {
+        DPRINTF(BAC,"[tid:%i] Fetch stall detected.\n",tid);
```
Style: needs spaces after commas.

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+    }
+
+    if (stalls[tid].bpu) {
+        DPRINTF(BAC,"[tid:%i] BPU stall detected.\n",tid);
```
Style: needs spaces after commas.

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+            bpu->squash(fromCommit->commitInfo[tid].doneSeqNum, tid);
+            if (fromCommit->commitInfo[tid].mispredictInst) {
+                DPRINTF(BAC, "[tid:%i] Squashing due to mispredict of "
+                        "non-control instruction: %s\n",tid,
```
Style: should have spaces after commas.

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+                                fromCommit->commitInfo[tid]
+                            .mispredictInst->pcState().instAddr()));
+            } else {
+                DPRINTF(BAC, "[tid:%i] Squashing due to "
```
Is this debug message correct? This is the `!fromCommit->commitInfo[tid].mispredictInst` case, so it's not a squash due to misprediction.

dhschall replied:
This is the else case for the `isControl` check

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+    // Check stalls
+    if (stalls[tid].drain) {
+        assert(cpu->isDraining());
+        DPRINTF(BAC,"[tid:%i] Drain stall detected.\n",tid);
```
Style: should have spaces after commas.

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+    if (checkStall(tid)) {
+        // return block(tid);
+        bacStatus[tid] = Blocked;
+        return false;
```
Is this correct, or should it be `true`? Looks like the status could have changed to `Blocked`.

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+
+    DPRINTF(BAC, "%s(tid:%i): FTQ sz: %i\n", tid, __func__, ftq->size(tid));
+
+    unsigned n_fts = ftq->size(tid);
```
This could probably be replaced with `if (ftq->size(tid) == 0) return;` for simplicity (or even `if (ftq->isEmpty(tid)) return;`, or just skipped as `forAllBackwards` does nothing for an empty FTQ).

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+    bool status_change = false;
+
+    std::list<ThreadID>::iterator threads = activeThreads->begin();
+    std::list<ThreadID>::iterator end = activeThreads->end();
```
Nit: I might have made this `const`, since it's the stopping point of the loop.

dhschall replied:
Is that correct if I change it to
```
for (ThreadID tid : *activeThreads) {
```

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+    Addr search_addr = cur_pc.instAddr();
+    Addr start_addr = search_addr;
+
+    // In each cycles a new fetch target is created starting with
```
Typo: cycles -> cycle

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+
+        // If its not a branch check if the maximum search width is reached.
+        // If yes stop searching.
+        if ((search_addr - start_addr) >= fetchTargetWidth) {
```
I don't think it matters for correctness in this implementation, but this seems to be off-by-one in terms of the number of bytes searched (because of the placement of the test). The logic would probably be easier with the test after the increment, or in the `while` loop test.

dhschall replied:
Really? I don't think so. I start with the first address check if a branch exist and if it exceeds the max width.
If not I increment and search the next address

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+
+        // Not a branch therefore we will continue the next FT at the
+        // next address
+        next_pc->set(cur_pc.instAddr() + minInstSize);
```
I might be wrong about this, but should the `next_pc` be the `cur_pc` plus the size of the actual instruction at the end of the search window, rather than `minInstSize`? In the fall-through case, if the last instruction of the FT is not `minInstSize` bytes long (for X86), or `minInstSize` does not match the instruction size for fixed-width encoding ISAs, then the next_pc could land on a non-instruction boundary. 

`next_pc` will then be used as the start address of the next FT. When the FT is consumed, is the start address assumed to be a valid instruction address?

`minInstSize` is documented as an optimisation (i.e. `minInstSize=1` should work for Arm and RISC-V), but using it in this way here means it must be set correctly.

dhschall replied:
You are right that this would mean that for Arm, RISC-V a fetch target would start within the previous instruction. I don't think it will cause a problem though as in the fetch unit we check if the current PC is within the range of the fetch target.

Do you have a suggestion how to make it better, more explicit?

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+    // Finally set the BPU PC to the next FT in the next cycle
+    set(cur_pc, *next_pc);
+
+    // ftq->printFTQ(tid);
```
Remove commented code?
(Another option would be to leave the call to `printFTQ` but guard it with a check for the `FTQ` debug flag.)

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+        ftq->lock(tid);
+    }
+
+    // 2. For complex instruction with more than one branches the history
```
Typo: "branches" -> "branch in"

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+    if (inst->isMicroop() && !inst->isLastMicroop() && (hist == nullptr)) {
+
+        DPRINTF(Branch, "No history for complex instruction found. \n");
+        stats.multiBranchInst++;
```
Does this over-count the complex instructions? From your explanation it sounds like it counts the complex instruction micro-ops, not the overall instructions. 
(Not sure if there is an easy way to fix this though.)

dhschall replied:
I just wanted to count how often this situation is happening

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+    // Assign the branch instruction instance its sequence number
+    // and push the history to the main history buffer.
+    hist->seqNum = seqNum;
+    bpu->predHist[tid].push_front(hist);
```
Nit: it would be nicer to do this sort of thing with an API / encapsulated interface, rather than friendship.

dhschall replied:
I agree. That is a leftover hack.

dhschall replied:
Thanks for the reminder

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+    : statistics::Group(cpu, "bac"),
+
+    ADD_STAT(idleCycles, statistics::units::Cycle::get(),
+            "Number of cycles BAC is idle. (PC invalid)"),
```
Style nit: indentation is slightly off on many of these lines.

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+             "Number of branches that fetch encountered which are not the "
+             "last uOp within a macrooperation. Jump to itself."),
+    ADD_STAT(branchMisspredict, statistics::units::Count::get(),
+            "Number of branches that BAC has predicted taken"),
```
Copy-paste of the wrong description?

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+    ADD_STAT(branchMisspredict, statistics::units::Count::get(),
+            "Number of branches that BAC has predicted taken"),
+    ADD_STAT(noBranchMisspredict, statistics::units::Count::get(),
+            "Number of branches that BAC has predicted taken"),
```
Copy-paste of the wrong description?

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+    ADD_STAT(preDecUpdate, statistics::units::Count::get(),
+            "Number of branches extracted from the predecoder"),
+    ADD_STAT(noHistType, statistics::units::Count::get(),
+            "Number and type of branches that where undetected by the BPU."),
```
Typo: where -> were

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1065 @@
+            "Number of branches squashed from decode"),
+    ADD_STAT(preDecUpdate, statistics::units::Count::get(),
+            "Number of branches extracted from the predecoder"),
+    ADD_STAT(noHistType, statistics::units::Count::get(),
```
Would something like `noHistByType`, or just `noHistCounts`, be a slightly clearer name for this stat?

rjc-arch commented on src/cpu/o3/cpu.cc:
```diff
@@ -111,7 +114,7 @@ CPU::CPU(const BaseO3CPUParams &params)
                   params.activity),
 
-      globalSeqNum(1),
+      globalSeqNum(1), globalFTSeqNum(1),
```
Nit: would this look neater on its own line?

dhschall replied:
I've updated the .clang-format file

rjc-arch commented on src/cpu/o3/cpu.hh:
```diff
@@ -520,6 +534,9 @@ class CPU : public BaseCPU
     InstSeqNum globalSeqNum;//[MaxThreads];
 
+    /** The global sequence number counter. */
+    FTSeqNum globalFTSeqNum;//[MaxThreads];
```
Could the commented part be removed?

rjc-arch commented on src/cpu/o3/fetch.cc:
```diff
@@ -243,6 +249,9 @@ Fetch::setTimeBuffer(TimeBuffer<TimeStruct> *time_buffer)
     fromCommit = timeBuffer->getWire(-commitToFetchDelay);
+
+    // Create a wire to send information to BAC
+    toBAC = timeBuffer->getWire(0);
```
Would it be useful to have a short comment here to explain the zero delay, as it breaks the pattern of the other wires above?
(Although you _have_ explained it in other comments.)

dhschall replied:
Is that ok?

rjc-arch replied:
Thanks - looks good.

rjc-arch commented on src/cpu/o3/fetch.cc:
```diff
@@ -251,6 +260,14 @@ Fetch::setActiveThreads(std::list<ThreadID> *at_ptr)
 }
 
+void
+Fetch::setBACandFTQPtr(BAC *bac_ptr, FTQ * ftq_ptr)
```
Style: inconsistent spacings around `*`.

rjc-arch commented on src/cpu/o3/fetch.cc:
```diff
@@ -489,49 +506,20 @@ Fetch::deactivateThread(ThreadID tid)
-        inst->setPredTarg(next_pc);
-        inst->setPredTaken(false);
+    if (!decoupledFrontEnd) return true;
+    // If the FTQ is empty wait unit its filled upis available.
```
Typo: -> "If the FTQ is empty, wait until it is available."

rjc-arch commented on src/cpu/o3/fetch.cc:
```diff
@@ -688,6 +676,34 @@ Fetch::finishTranslation(const Fault &fault, const RequestPtr &mem_req)
+
+void
+Fetch::squashFromDecode(const PCStateBase &new_pc, const DynInstPtr squashInst,
+        const InstSeqNum seq_num, ThreadID tid)
```
Style: indentation doesn't match the style guide (although it is consistent with other parts of this file).

rjc-arch commented on src/cpu/o3/fetch.cc:
```diff
@@ -688,6 +676,34 @@ Fetch::finishTranslation(const Fault &fault, const RequestPtr &mem_req)
+
+void
+Fetch::squash(const PCStateBase &new_pc, const InstSeqNum seq_num,
+        DynInstPtr squashInst, ThreadID tid)
```
Style: indentation doesn't match the style guide (although it is consistent with other parts of this file).

rjc-arch commented on src/cpu/o3/fetch.cc:
```diff
@@ -688,6 +676,34 @@ Fetch::finishTranslation(const Fault &fault, const RequestPtr &mem_req)
+
+
+void
+Fetch::squash(const PCStateBase &new_pc, const InstSeqNum seq_num,
```
Should this be re-named `squashFromCommit` to match `squashFromDecode` above?

rjc-arch commented on src/cpu/o3/fetch.cc:
```diff
@@ -740,17 +756,16 @@ Fetch::doSquash(const PCStateBase &new_pc, const DynInstPtr squashInst,
-    DPRINTF(Fetch, "[tid:%i] Squashing from decode.\n", tid);
-
-    doSquash(new_pc, squashInst, tid);
+    DPRINTF(Fetch,"[tid:%i] Resteer BAC to PC: %s\n",tid, new_pc);
```
Style: needs spaces after commas.

rjc-arch commented on src/cpu/o3/fetch.cc:
```diff
@@ -1000,16 +972,28 @@ Fetch::checkSignalsAndUpdate(ThreadID tid)
         // squash this cycle.
-        DPRINTF(Fetch, "[tid:%i] Done squashing, switching to running.\n",
-                tid);
+        // With a decoupled front-end we can only to running if the FTQ
```
Typo: "we can only to running" -> "we can only switch to running"

rjc-arch commented on src/cpu/o3/fetch.cc:
```diff
@@ -1000,16 +972,28 @@ Fetch::checkSignalsAndUpdate(ThreadID tid)
 
+    // Check if the FTQ in not empty anymore.
+    if (fetchStatus[tid] == FTQEmpty && !ftq->isEmpty(tid)) {
+        DPRINTF(Fetch, "[tid:%i] FTQ is refilled -> running\n", tid);
```
Nit: the wording of this message could be more consistent with the ones above.

rjc-arch commented on src/cpu/o3/fetch.cc:
```diff
@@ -1087,16 +1071,39 @@ Fetch::fetch(bool &status_change)
+    // Check if the FTQ is ready and process the tail fetch target.
+    // In the non decoupled front-end ftqReady() will always return true;
+    if (!ftqReady(tid, status_change)) {
+        DPRINTF(Fetch, "FTQ not ready [tid:%i]\n", tid);
```
Nit: likewise, the `[tid:%i]` normally comes at the beginning of the debug messages in other places.

rjc-arch commented on src/cpu/o3/fetch.cc:
```diff
@@ -1087,16 +1071,39 @@ Fetch::fetch(bool &status_change)
         return;
     }
 
+    // Check if the FTQ is ready and process the tail fetch target.
```
Is this comment correct? It seems to explain the internals of `ftqReady`, and that looks like it examines (rather than processes) the head of the FTQ, not the tail.

rjc-arch commented on src/cpu/o3/fetch.cc:
```diff
@@ -1087,16 +1071,39 @@ Fetch::fetch(bool &status_change)
+    FetchTargetPtr curFT = ftq->readHead(tid);
+
+    if (decoupledFrontEnd) { // #ifdef FDIP
+        assert(ftqReady(tid,status_change));
```
Style: needs space after comma.

rjc-arch commented on src/cpu/o3/fetch.cc:
```diff
@@ -1304,8 +1337,20 @@ Fetch::fetch(bool &status_change)
                 "for this cycle.\n", tid);
     } else if (blkOffset >= fetchBufferSize) {
-        DPRINTF(Fetch, "[tid:%i] Done fetching, reached the end of the"
+        DPRINTF(Fetch, "[tid:%i] Done fetching, reached the end "
```
Typo: I think the "of the" is still needed.

rjc-arch commented on src/cpu/o3/fetch.cc:
```diff
@@ -1304,8 +1337,20 @@ Fetch::fetch(bool &status_change)
+        DPRINTF(Fetch, "[tid:%i] Done fetching, reached the end "
                 "fetch buffer.\n", tid);
+    } else if (decoupledFrontEnd && !curFT) {
+        DPRINTF(Fetch, "[tid:%i] Done fetching, reached end of fetch "
```
Typo: -> end of the fetch target.

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -727,6 +727,10 @@ BAC::generateFetchTargets(ThreadID tid, bool &status_change)
     // This could be circumvented by using not only the PC but also the
     // microPC to make predictions. However, since such instructions are
     // rare this is not implemented.
+    // However, arm does not micro code branches and correspondingly does
```
Nit: if it's not too hard to do, you could move this chunk to the previous commit in the PR. It looks like an addition that was added to the wrong commit.

rjc-arch commented on src/mem/cache/prefetch/base.hh:
```diff
@@ -385,6 +386,8 @@ class Base : public ClockedObject
     virtual void
     setParentInfo(System *sys, ProbeManager *pm, unsigned blk_size);
 
+    virtual void setCache(BaseCache *cache) {}
```
If this does not get replaced with the `CacheAccessor` interface, would it be worth adding a documentation comment here that the `Base` prefetcher does not have a cache pointer, and this method is used to provide the service to prefetchers that do need one. I think I do agree overall that it's better not to add a `BaseCache* cache = nullptr;` member to `Base`.

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+{
+    /* Create a prefetch memory request */
+    RequestPtr req = nullptr;
+    Flags flags = Request::INST_FETCH|Request::PREFETCH;
```
Style: there should be whitespace around the `|`.

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+
+    if (virtual_addr) {
+        // The address is virtual -> we need translate first
+        req = std::make_shared<Request>(
```
Could this use the `createPrefetchRequest` helper method?

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+{
+    RequestPtr req = std::make_shared<Request>(
+            vaddr, blkSize, 0, requestorId, vaddr, 0);
+    req->setFlags(Request::PREFETCH);
```
`createPrefetchPacket` seems to replicate some of this functionality, and also uses slightly different flags (it also sets `Request::INST_FETCH`).
Should the flags be the same and could `createPrefetchRequest` be re-used in `createPrefetchPacket`.

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+
+        // Translate the address from virtual to physical using
+        // the functional translation function.
+        // Note: This of course a hack and is not how it is in a real system.
```
Typos: This of course -> This is of course
underestimate -> underestimates

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+
+        if (cacheSnoop &&
+            (cache->inCache(it->req->getPaddr(), it->req->isSecure())
+         || (cache->inMissQueue(it->req->getPaddr(), it->req->isSecure())))) {
```
Style: indentation is slightly off here.

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+
+
+RequestPtr
+FetchDirectedPrefetcher::createPrefetchRequest(Addr vaddr)
```
Does this method get called from anywhere? The code compiles without it.

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+
+
+PacketPtr
+FetchDirectedPrefetcher::createPrefetchPacket(Addr addr, bool virtual_addr)
```
Does this method get called from anywhere? The code compiles without it.

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+
+
+void
+FetchDirectedPrefetcher::notifyPfAddr(Addr addr, bool virtual_addr)
```
This method seems to assume `virtual_addr` is always true (i.e. it always translates the address).

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+{
+    Flags flags = Request::INST_FETCH|Request::PREFETCH;
+    req = std::make_shared<Request>(
+            addr, owner.blkSize, flags, owner.requestorId, addr, 0);
```
`ContextID` is hard coded to zero here, then used in `PrefetchRequest::startTranslation` to get a `ThreadContext` for the translation. Does that indicate the code will not work for `numThreads > 1`? Could an actual ContextID be used?

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+}
+
+
+void
```
Style nit: it would feel more natural if this method were below the `translateFunctional` method.

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+{
+    auto it = translationq.begin();
+    while (it != translationq.end()) {
+        if (&(*it) == pfr) {
```
This test seems quite fragile. It works for `Request`s in a `std::list`, but would break if the container is switched out for an alternative that copies or moves its contents (e.g. a `std::vector` or `std::deque`). I think it would be better to implement a comparison operator in the `PrefetchRequest` class to test for object equality rather than object identity. (Happy for that to be as a future improvement, though. I don't think it is _incorrect_ as it stands.)

rjc-arch replied:
Update: it looks like `PrefetchRequest`s can't be copied or moved as currently implemented anyway (see comment below), so perhaps this is not so much of an issue.

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+    auto tc = system->threads[req->contextId()];
+
+    DPRINTF(HWPrefetch, "%s Try trans of pc %#x\n",
+                                mmu->name(), req->getVaddr());
```
Style: indentation here is probably not correct.

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+
+
+bool
+FetchDirectedPrefetcher::translateFunctional(RequestPtr req)
```
Does this method get called from anywhere? The code compiles without it.

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+    Flags flags = Request::INST_FETCH|Request::PREFETCH;
+    req = std::make_shared<Request>(
+            addr, owner.blkSize, flags, owner.requestorId, addr, 0);
+    req->setFlags(Request::PREFETCH);
```
Is this line redundant? `Request::PREFETCH` is already part of `flags`.

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+void
+FetchDirectedPrefetcher::PrefetchRequest::createRequest()
+{
+    Flags flags = Request::INST_FETCH|Request::PREFETCH;
```
Style: needs whitespace around `|`.

rjc-arch commented on src/mem/cache/prefetch/fdp.hh:
```diff
@@ -0,0 +1,209 @@
+    /** The prefetch queue entry objects */
+    struct PrefetchRequest : public BaseMMU::Translation
+    {
+        // PrefetchRequest(FetchDirectedPrefetcher& _owner,
```
Style: remove commented code.

rjc-arch commented on src/mem/cache/prefetch/fdp.hh:
```diff
@@ -0,0 +1,209 @@
+        PrefetchRequest(FetchDirectedPrefetcher& _owner, uint64_t _addr);
+
+        /** Owner of the packet */
+        FetchDirectedPrefetcher& owner;
```
It might be worth explicitly marking this class as unmoveable and uncopyable, since the `FetchDirectedPrefetcher&` member effectively makes it so. It also precludes the use of some containers for the `ftq` and `translationq` (the current `std::list` is fine, but a `std::vector` or `std::deque`couldn't be used as things currently stand).

Alternatively, you could consider using a pointer here, instead of a reference, for added future flexibility.

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+}
+
+void
+FetchDirectedPrefetcher::PrefetchRequest::createPkt(Tick t)
```
Suggestion: would it be clearer to separate this into two calls, one to create the packet, and a separate method set the `readyTime`?

The current version allows simpler code (one call instead of two), but mixes two actions. Setting the readyTime isn't really part of crating a packet, and it's conceivable you might want to change the readyTime without re-allocating a new packet.

This is subjective, so feel free to disagree with me on this point.

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+{
+    assert(owner.mmu != nullptr);
+    auto tc = owner.system->threads[req->contextId()];
+    // DPRINTF(HWPrefetch, "%s Try translation of pc %#x\n",
```
Style: remove or uncomment this line?

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+    Base::regProbeListeners();
+
+    if (cpu == nullptr) {
+        warn("No CPU to listen from registered\n");
```
It might be helpful to add a little more context to this message, as it will be produced during CPU initialisation. (e.g. add the prefix "FetchDirectedPrefetcher: " to the message.)

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+    ADD_STAT(fdipInsertions, statistics::units::Count::get(),
+            "Number of notifications from an insertion in the FTQ"),
+    ADD_STAT(pfIdentified, statistics::units::Count::get(),
+            "number of prefetches identified."),
```
Style nit: inconsistent capitalisation of these documentation strings. (Personally I prefer capitalised, but I think  consistency is the most important aspect.)

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,337 @@
+    ADD_STAT(pfPacketsCreated, statistics::units::Count::get(),
+            "number of prefetch packets created"),
+    ADD_STAT(pfCandidatesAdded, statistics::units::Count::get(),
+            "Number of perfetch candidates added to the prefetch queue"),
```
Typo: perfetch -> prefetch

rjc-arch commented on src/mem/cache/prefetch/fdp.hh:
```diff
@@ -0,0 +1,209 @@
+    /** Notify functions are not used by this prefetcher. */
+    void notify(const CacheAccessProbeArg &acc, const PrefetchInfo &pfi)
+        override {};
+    // void notifyFill(const PacketPtr &pkt) override{};
```
Style: remove commented code.

rjc-arch commented on src/mem/cache/prefetch/fdp.hh:
```diff
@@ -0,0 +1,209 @@
+    BaseCache *cache;
+
+    /** For testing purposes */
+    const bool transFunctional;
```
This doesn't seem to be used at the moment. Could it be removed?

rjc-arch commented on src/mem/cache/prefetch/fdp.hh:
```diff
@@ -0,0 +1,209 @@
+        /** Owner of the packet */
+        FetchDirectedPrefetcher& owner;
+
+        /** The virtual address. Used to scan for redundand prefetches.*/
```
Typo: redundand -> redundant

rjc-arch commented on src/mem/cache/prefetch/fdp.hh:
```diff
@@ -0,0 +1,209 @@
+        bool operator==(const int& a) const {
+            return this->addr == a;
+        }
+
```
Style nit: Move this blank line above `operator==`.

rjc-arch commented on src/mem/cache/prefetch/fdp.hh:
```diff
@@ -0,0 +1,209 @@
+
+
+        void finish(const Fault &fault, const RequestPtr &req,
+                      ThreadContext *tc, BaseMMU::Mode mode) override;
```
Style nit: incorrect indentation.

rjc-arch commented on src/mem/cache/prefetch/fdp.hh:
```diff
@@ -0,0 +1,209 @@
+        */
+        void startTranslation();
+
+        void markDelayed() override{}
```
Style nit: needs whitespace after `override`.

rjc-arch commented on src/mem/cache/prefetch/fdp.hh:
```diff
@@ -0,0 +1,209 @@
+     * @param addr is the start address of the fetch target
+     * @param va is true if the address is a virtual address
+     * */
+    void notifyPfAddr(Addr addr, bool va=false);
```
Style nit: the name of the second parameter in this declaration does not match the definition. (`va` here and in the doc-comment; `virtual_addr` in the definition.)

rjc-arch commented on src/mem/cache/prefetch/fdp.hh:
```diff
@@ -0,0 +1,209 @@
+
+    void translationComplete(PrefetchRequest* pf_req, const bool failed);
+
+    /** Performs a functional translation of the incomming packet by useing
```
Typo: useing -> using

rjc-arch commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,333 @@
+    # Any resource specified in this file will be automatically retrieved.
+    # At the time of writing, this file is a WIP and does not contain all
+    # resources. Jira ticket: https://gem5.atlassian.net/browse/GEM5-1096
+    obtain_resource(workloads[args.workload][args.isa])
```
Optional: this could be re-worked to allow the local path of an ELF to be specified, then fall back a set of off-the-shelf workloads from the gem5 resources (as currenlty implemented). That would allow the user to easily run local binaries as well. (The `hello` workload is great for checking everything is working, but probably too small to properly evaluate this new feature.)

rjc-arch commented on src/cpu/o3/cpu.hh:
```diff
@@ -520,6 +534,9 @@ class CPU : public BaseCPU
     /** The global sequence number counter. */
     InstSeqNum globalSeqNum;//[MaxThreads];
```
Could the commented part be removed?

rjc-arch reviewed (changes requested):
Hi @dhschall,

I’ve now completed a detailed review — please see my inline comments below. Overall, it’s looking very good. As you requested, I took a thorough approach and flagged even minor issues, so please accept my apologies in advance for the volume of comments!

That said, I think this is a very well-implemented PR; the code is clean and readable, and it integrates well with the existing structure.

Unfortunately, I wasn’t able to identify any obvious bugs that might explain the performance differences we’ve been seeing.

rjc-arch commented:
Hi @powerjg,
 
> One thing we should check is that when the decoupled front end is turned off the performance should be about the same as before.

From initial testing, I can confirm that the the simulated IPC is unchanged from the baseline when the decoupled front end is disabled.

At the moment, we are not seeing the expected IPC uplift with the feature enabled, so we still have some more investigation ahead of us before we're ready to merge this PR.

powerjg commented:
Thanks for the effort here, @rjc-arch and @dhschall!!

dhschall commented:
Thanks a lot @rjc-arch and @rjc-arch for the great feedback! I'm was travelling the last days but once I'm back I'll address all the comments!

dhschall committed:
tests: Add tests for decoupled front-end

dhschall commented:
Hi @powerjg and @rjc-arch,
I have now reworked the PR and tried to integrate all your feedback. The biggest changes are:
- I run the clang-formatter on all commits to fix all code formatting issues.
- I significantly reworked and cleaned up the `fdp.hh/.cc` files to remove all redundant code and fixed some issues with setting the correct context ID. These files may need another code review.
- I also removed the `setCache` function and made the cache snooping optional. This way the FDP prefetcher should also be compatible with Ruby
- I added new CI tests for the decoupled front-end. These tests run for X86, Arm and RISC-V.

Also, @rjc-arch ran some tests now and it looked promising.
Since there were so many comments I might have overlooked something. Let me know if thats the case

rjc-arch commented on src/cpu/o3/bac.cc:
```diff
@@ -0,0 +1,1032 @@
+        stalls[i] = {false, false, false};
+    }
+
+    assert(bpu != nullptr);
```
Optional: it might be more convenient to make this an assert with a message or another type of gem5 message to help the user debug their config.

rjc-arch commented on src/cpu/o3/bac.hh:
```diff
@@ -0,0 +1,472 @@
+    /** Number of threads. */
+    const ThreadID numThreads;
+
+    /*Max number of FT added to the FTQ per Cycle */
```
Nit: these two comments should have an extra space after the `/*`, and perhaps a blank line between for consistency.

rjc-arch commented on src/cpu/o3/cpu.hh:
```diff
@@ -526,7 +543,10 @@ class CPU : public BaseCPU
-    InstSeqNum globalSeqNum;//[MaxThreads];
+    InstSeqNum globalSeqNum;
+
+    /** The global sequence number counter. */
```
Nit: should this be "The global FT sequence number counter."?

dhschall replied:
Yes thanks!

rjc-arch commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,358 @@
+     * Requires squash to recover. */
+    void invalidate(ThreadID tid);
+
+    /** Returns if the FTQ is in a ready state and its save to consmume
```
Typo: should be 'save' -> 'safe'.

rjc-arch commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,358 @@
+    /** Whether the exit instruction is a branch */
+    bool is_branch;
+
+    /** If the exit branch is taken */
```
Nit: this should probably be 'If the exit branch is predicted taken'.

rjc-arch commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,358 @@
+    /** Check if the FTQ is locked. */
+    bool isLocked(ThreadID tid);
+
+    /** Interates forward over all fetch targets in the FTQ from head/front to
```
Typo: should be 'Iterates'.

rjc-arch commented on src/cpu/o3/ftq.hh:
```diff
@@ -0,0 +1,358 @@
+     * tail/back and applies a given function. */
+    void forAllForward(ThreadID tid, std::function<void(FetchTargetPtr &)> f);
+
+    /** Interates backward over all fetch targets in the FTQ from tail/back to
```
Typo: should be 'Iterates'.

rjc-arch commented on src/mem/cache/prefetch/Prefetcher.py:
```diff
@@ -727,3 +728,41 @@ def listenFromProbeRetiredInstructions(self, simObj):
+    type = "FetchDirectedPrefetcher"
+    cxx_class = "gem5::prefetch::FetchDirectedPrefetcher"
+    cxx_header = "mem/cache/prefetch/fdp.hh"
+    cxx_exports = [PyBindMethod("setCache")]
```
Thanks for this improvement, @dhschall.

rjc-arch commented on src/mem/cache/prefetch/Prefetcher.py:
```diff
@@ -727,3 +728,41 @@ def listenFromProbeRetiredInstructions(self, simObj):
+    )
+    squash_prefetches = Param.Bool(
+        True,
+        "Squash the prefetch queue in case a fetch target is removed from "
```
This comment seems a little misleading. 
Rather than squashing the whole queue, the parameter causes the prefetches associated with the removed FT to be removed from the prefetch queue.

dhschall replied:
Makes sense! Thanks!

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,308 @@
+namespace gem5
+{
+
+GEM5_DEPRECATED_NAMESPACE(Prefetcher, prefetch);
```
I can't find any other deprecation warnings for the `Prefetcher` namespace in the current `develop` branch. It looks like `Prefetch` has finished it's deprecation period now, so this line could be removed.

rjc-arch commented on src/mem/cache/prefetch/fdp.cc:
```diff
@@ -0,0 +1,308 @@
+      ADD_STAT(pfIdentified, statistics::units::Count::get(),
+               "Number of prefetches identified."),
+      ADD_STAT(pfSquashed, statistics::units::Count::get(),
+               "Number of prefetches identified."),
```
This should be "Number of prefetches squashed.".

rjc-arch commented on src/mem/cache/prefetch/fdp.hh:
```diff
@@ -0,0 +1,206 @@
+
+struct FetchDirectedPrefetcherParams;
+
+GEM5_DEPRECATED_NAMESPACE(Prefetcher, prefetch);
```
As mentioned above, it looks like Prefetch has finished it's deprecation period now, so this line could be removed.

rjc-arch commented on configs/example/gem5_library/fdp-hello.py:
This example looks really good.

rjc-arch reviewed (approved):
Hi @dhschall,

This is looking very good now. I have commented on a few very minor/cosmetic outstanding issues. Most are just typos, but a couple are user-facing (documentation comments or stats labels). You could address them now, but I also think it would be reasonable to merge this PR as it currently stands, and fix the last few errata in a separate PR if you prefer. 

I'm adding my approval now as I think this PR could be merged as it stands. 

Thanks for all your hard work on this feature, @dhschall!

rjc-arch commented:
Additionally, @dhschall and I have performed extensive performance testing of this feature.

- The latest revision performs identically to the previous version reviewed above (when configured equivalently; David has added some new configuration options and changed the default options slightly).
- The new FDP does not degrade simulated performance when the Decoupled Front-End is disabled (which is the case by default).
- When the Decoupled Front-End and Fetch-Directed Prefetcher are enabled, we observe IPC uplifts roughly in line with those expected from the academic literature for workloads with high L1I Cache miss rates.

dhschall commented:
Thanks @rjc-arch for the second code review and your huge effort in testing these patches!!!. I've incorporated all your suggestions.

@powerjg I removed your main concerns with the cache interface. So it should also work with ruby.
Let me know if I missed something. Otherwise its now up to you to decide on the last three digit PR 😅

powerjg commented on configs/example/gem5_library/fdp-hello.py:
```diff
@@ -0,0 +1,339 @@
+# 2. Decoupled front-end ------------------------------------------------
+# Next setup the decoupled front-end. Its implemented in the O3 core.
+# Create the processor with one core
+processor = SimpleProcessor(
```
As a future PR, we should create a processor & cache hierarchy which has fdp enabled in the stdlib. We'll have to think a little about the interface for this since it requires coordination between the cache hierarchy and the processor.

dhschall replied:
I've now created a draft for this #2595 
The [commit](https://github.com/gem5/gem5/pull/2595/commits/e7b2383948dfaa10e55a917d6998c41986be09f9) adds two components: Processor and CacheHierarchy with decoupling preconfigured.

I think this should be independent of this PR.

powerjg commented on src/cpu/pred/bpred_unit.hh:
```diff
@@ -407,6 +400,9 @@ class BPredUnit : public SimObject
     void updateBTB(ThreadID tid, PredictorHistory *&bpu_history);
 
+  private:
+    typedef std::deque<PredictorHistory *> History;
```
I believe this should be lowercase `history`. See [Naming](https://www.gem5.org/documentation/general_docs/development/coding_style/#naming)

dhschall replied:
Really? Also for type definitions? Its not a variable declaration.

dhschall replied:
On the other hand, I think the `History` type is just used once, so I'll remove it altogether.

powerjg replied:
Ok. Sorry, I thought it was a variable. I missed the typedef.

powerjg reviewed (commented):
One tiny nit. Overall, this is great. I'm excited to get it in!

Is there a citation for this? I know it's a ton of effort, and you should get some credit!

dhschall commented:
> Is there a citation for this? I know it's a ton of effort, and you should get some credit!
I developed it for this work
```
@inproceedings{10.1145/3613424.3614258,
  author    = {Schall, David and
               Sandberg, Andreas and
               Grot, Boris},
  title     = {Warming Up a Cold Front-End with Ignite},
  year      = {2023},
  publisher = {Association for Computing Machinery},
  address   = {Toronto, ON, Canada},
  doi       = {10.1145/3613424.3614258},
  booktitle = {Proceedings of the 56th Annual IEEE/ACM International Symposium on Microarchitecture (MICRO '23)},
  series    = {MICRO'23}
}
```
I'll add it to the `BaseCPU` with the `add_citation`

dhschall commented on src/cpu/o3/O3CPU.py:
```diff
@@ -52,3 +53,20 @@
         from m5.objects.X86CPU import X86O3CPU as O3CPU
 
     DerivO3CPU = O3CPU
+
```
@powerjg I've added the citation now here and to the prefetcher. Is that correct?

dhschall commented:
@powerjg I've added the citation to the BaseO3CPU and the prefetcher. And also an initial draft to support it in the standard library (but I think it should be a separate PR).
What do you think, is it ok as it is?

powerjg reviewed (approved):
this looks great to me. Let's merge it and work out any final kinks that come up. Let's pay close attention to nightly and weekly tests after this is merged. (Unfortunately, the nightly/weekly have been failing for 3 weeks due to the PCI changes. I *think* they will pass today.)
